### ISSUE TITLE
Colored itemstacks are broken on create or dig, if palette index is 0

### ISSUE THREAD

**Issue by VanessaE** [minetest/minetest#10687]
##### Minetest version

"5.4.0-dev", at commit cfaef5b1cfc8ec7463a2e8aec420bfc3a434f9c1

##### OS / Hardware

Not related.

##### Summary

I was trying to address an issue in Unified Dyes ([as reported on the forum](https://forum.minetest.net/viewtopic.php?p=385051#p385051)), but it turns out to be an engine problem, though I haven't the faintest clue how long it's been the case.

Originally my code did this, to create a colored itemstack:

```lua
function unifieddyes.make_colored_itemstack(item, palette, color)
	local paletteidx = unifieddyes.getpaletteidx(color, palette)
	local stack = ItemStack(item)
	stack:get_meta():set_int("palette_index", paletteidx)
	return stack:to_string(),paletteidx
end
```

While the values passed into the function from elsewhere are correct, and `unifieddyes.getpaletteidx()` is returning the correct results (both verified by a bunch of print statements), any time the index was `0`, the final item string had no color information, leading me to conclude that an integer metadata key with a value of `0` is treated internally by the engine as equivalent to `nil` i.e. non-existent.

This matters because, while other mods might use `0` for white or perhaps some "neutral" color, Unified Dyes puts "faint red" there.

I figured, okay, I remember seeing that "0 = `nil`" thing before in another mod, and there is a proper API call for this now anyway, so my code's just outdated; a hold-over from before that function was added.

I altered my code thus:

```lua
function unifieddyes.make_colored_itemstack(item, palette, color)
	local paletteidx = unifieddyes.getpaletteidx(color, palette)
	return minetest.itemstring_with_palette(item, paletteidx), paletteidx
end
```

But that failed too, in exactly the same way.  The engine is returning just the original unmodified item string when the palette index is `0`. That is,  `minetest.itemstring_with_palette("coolmod:my_awesome_node", 0)` just returns `coolmod:my_awesome_node`.

On the other hand, passing the same item but with a non-zero index, works.  For example, this:
`minetest.itemstring_with_palette("coolmod:my_awesome_node", 24)`
... returns this:
`coolmod:my_awesome_node 1 0 \"\\u0001palette_index\\u000224\\u0003\"`

On a whim, I tried the, `minetest.itemstring_with_color()` function, and it seems to treat a color index of `0` properly.  This:
`minetest.itemstring_with_color("coolmod:my_awesome_node", 0)`
... returns this:
`coolmod:my_awesome_node 1 0 \"\\u0001color\\u00020\\u0003\"`

Which I suppose is the expected output.  So if one were to swap `color` for `palette_index` via `string.gsub()`, this could be used to to work around the craft-place issue.

However, this isn't enough by itself:  when you dig the now-correctly-formed node, it just turns white again.  I.e. if the palette index was `0`, even though the node looks right in the world, the color info is stripped-out when the node passes through `minetest.node_dig()`, so the item turns white when it reaches the inventory, so you have to work around this loss of color when digging, too.

**Comment by sfan5** [comment]
Broken by #8770 and fixed by #10183.
You are literally a single commit behind the fix.

**Comment by VanessaE** [comment]
Well, shit.  Ain't that the story of my life :smile:
